### ISSUE TITLE
Uses much memory on Chrome possibly because long promise chain uses memory a lot on Chrome

### ISSUE THREAD

**Issue by tyoshino** [domenic/streams-demo#4]
I think we discussed similar topic in the context of reviewing some example written by you for the Streams spec.

We got this report about fetch()'s memory usage. http://crbug.com/473062

We see ~100MB memory usage that doesn't decrease even we force GC. It seems it's for the promise chain.

If I changed the code to remove the chain, I saw only up to ~22MB memory usage on Chromium task manager.

Adding @yutakahirano to see if this is a problem to be solved by promises perspective or not.

```
    function pump(reader) {
        return new Promise(function (resolve, reject) {
            function pumpStep() {
                reader.read().then(function (result) {
                    if (result.done) {
                        progressText.textContent = `All done! (${soFar} bytes total)`;
                        resolve();
                        return;
                    }

                    ...

                    if (found !== -1) {
                        progressText.textContent = `Found it! At position ${soFar - chunk.byteLength + found}.`;

                        resolve(reader.cancel());
                    } else {
                        pumpStep(reader);
                    }
                });
            }
            pumpStep();
        });
    }
```

**Comment by wanderview** [comment]
Isn't chaining one of the main things Promises bring in terms of ergonomics?  Can we explain why chaining is increasing memory usage here?

**Comment by ithinkihaveacat** [comment]
My speculation in https://github.com/whatwg/fetch/issues/33 is that it's because two (?) stack frames are needed per read() chunk. (In this case tail call optimization could clean things up, but I don't think Chrome has it.) Not sure how to verify this, though.

**Comment by domenic** [comment]
This seems related ... https://github.com/promises-aplus/promises-spec/issues/179 there might be minor promise spec fixes here in terms of how their internal data structure is represented.

That crbug 403s for me.

**Comment by domenic** [comment]
Also we definitely should be able to use chaining here. If we can't that's either a spec bug (as speculated in my previous post) or just an implementation bug.

**Comment by yutakahirano** [comment]
Does that 70MB come from Promise objects? Or does it come from data attached to them?

I run the following code on  Chrome 41.0.2272.118 m (64-bit), but no memory blow-up is observed.

```
var counter = 0;
var k = 0;
function read() {
  var resolve;
  setTimeout(function() {
    ++counter;
    var buffer = new Uint8Array(1024 * 1024 * 4);
    buffer[counter % buffer.byteLength] = counter;
    resolve({done: false, value: buffer});
  }, 10);
  return new Promise(function(r) { resolve = r; });
}

function pump() {
  return read().then(function(r) {
    if (r.done) {
      console.log('done!');
      return;
    }
    // some computation
    for (var i = 0; i < r.value.byteLength; ++i) {
      k += r.value[i];
    }
    return pump();
  });
}

var p = pump();

setInterval(function() {
  console.log('k = ' + k);
}, 1000);
```

**Comment by tyoshino** [comment]
OK. So, this should be discussed with V8 team.

Sorry if it's just a Chromium/V8 bug. Please point this issue or something else better when you get similar report so that we don't give impression to users that `fetch()` is memory heavy.

I agree with wanderview's point basically. But I tend to write like https://github.com/domenic/streams-demo/issues/4#issue-68686793 for pumping repeated many times as chaining doesn't improve readability of the code so much for this case. Let me explain my feeling. Yes, it's just feeling.

When we enumerate async processing depending on some other async processing, promises help a lot for readability of the code:

```
foo().then((x) => {
  ...
}).then((transformedX) => {
  ...
}).then((transformedTransformedX) => {
  ...
```

Even if they don't want to pass the transformed data, it just helps readability.

For repeated pumping,

```
function pump() {
  if (...) {
    return deliverable;
  } else {
    return async().then(pump);
  }
}
```

```
function pump() {
  new Promise((resolve, reject) => {
    function sub() {
      if (...) {
        resolve(deliverable);
      } else {
        async().then(sub).catch(reject);
      }
    }
    sub();
  });
}
```

The former is a little simplified, but the readability improvement is not so big at least for me compared to the improvement for non-pumping case. I rather feel a bit uncomfortable with unnecessary use of chain for propagating deliverable to the top promise.

**Comment by domenic** [comment]
> OK. So, this should be discussed with V8 team.

Maybe. I think we want to produce a repro that's independent of fetch() first. @yutakahirano's post is interesting in that regard.

I wonder if it's possible for chaining to "leak" given ScriptPromises created by Blink, in scenarios where it doesn't leak when just using Promises created from JS directly? Just guessing here; it's been a while since I read @yutakahirano's ScriptPromise design doc.

**Comment by tyoshino** [comment]
Added `.catch(reject)`. The last half of the previous comment is explaining my programming preference. Not pushing it to everyone.

> Maybe. I think we want to produce a repro that's independent of fetch() first. @yutakahirano's post is interesting in that regard.

Got it. Investigating.

**Comment by tyoshino** [comment]
I created a reduced version of Yutaka's test with which we can create a longer chain quickly. Review appreciated. This increases memory usage but different form the Blink's fetch(), it gets collected when I force GC.

```
var counter = 0;

function chain() {
  ++counter;

  if (counter % 1000 !== 0) {
    return Promise.resolve().then(chain);
  }

  return new Promise(function(r) {
    setTimeout(function() {
      r(chain());
    }, 10);
  });
}

chain().then(function() {
  console.log('done');
});

setInterval(function() {
  console.log('counter = ' + counter);
}, 1000);
```

**Comment by domenic** [comment]
If it's collected when you force GC then I am not worried. I am still looking for a repro of an actual leak. 

The linked crbug (no idea why it's private, but now I have access when I use my @google account) says it only repros on OS X, not on Windows or Linux. So this is seeming more and more like a Chrome bug and not related to promises in general.

**Comment by ithinkihaveacat** [comment]
@tyoshino How do you force GC?

Memory use slowly grows on Node and Chrome (on OS X) with the code below, which attempts to trigger GC via:
- An occasional 1000ms pause.
- A forced called to `gc()` if available. (Running the code `node --expose-gc` makes this function available, but I'm not sure what it's actually doing.)

``` js
var count = 0;
var total = 0;

function read() {
  return new Promise(function (resolve) {
    setTimeout(function() {
      resolve(Math.random());
    }, count % 10000 ? 0 : 1000); // 1000ms pause to let gc kick in
  });
}

function pump() {
  count++;
  if (count % 10000 === 0 && typeof gc === 'function') { gc(); } // force gc if available
  return read().then(function (i) {
    total += i; // do some computation with i
    return pump();
  });
}

pump();

// Comment out on Chrome (process.memoryUsage() not available there)
setInterval(function() {
  console.log({ rss: (process.memoryUsage()).rss, count: count, average: total / count });
}, 1000);
```

Re @yutakahirano's [code above](https://github.com/domenic/streams-demo/issues/4#issuecomment-93442175), it seems memory usage does grows if `var buffer = new Uint8Array(...)` is moved out of `read()` and made a global. I guess those allocations are collected by GC, and over short periods of time, they dominate the memory profile, making it look like all memory is being reclaimed. But I think over time (e.g. 100k calls to `pump()`) memory usage does grow.

Re @tyoshino's [reduced version](https://github.com/domenic/streams-demo/issues/4#issuecomment-93465001), it does grow very fast if:

``` js
return Promise.resolve().then(chain);
```

is replaced with

``` js
return new Promise(function (resolve) {
  resolve(chain());
});
```

which I think should be equivalent (?). Not sure why they're so different.

**Comment by tyoshino** [comment]
> Re @tyoshino's reduced version, it does grow very fast if:

The Promise constructor runs the given function synchronously while `Promise.resolve().then(chain)` enqueues a microtask to run `chain` with undefined as an argument.

**Comment by tyoshino** [comment]
> @tyoshino How do you force GC?

Please open the DevTools and open the "Timeline" pane. Clicking the trash can icon forces GC.

**Comment by yutakahirano** [comment]
> Re @yutakahirano's code above, it seems memory usage does grows if var buffer = new Uint8Array(...) is moved out of read() and made a global. I guess those allocations are collected by GC, and over short periods of time, they dominate the memory profile, making it look like all memory is being reclaimed. But I think over time (e.g. 100k calls to pump()) memory usage does grow.

So, my question is, is the original problem caused by a large number of promise objects, or by Uint8Array objects? How much memory is consumed by 100k pump() calls in your environment? Domenic's demo calls pump() 60659 times in my environment (Linux x64), and the number of calls may be platform-dependent.

**Comment by yutakahirano** [comment]
I ran the script without Uint8Array, and succeeded to reproduce the memory blow-up. I can reproduce the issue with the standalone d8 (V8 shell).

**Comment by ithinkihaveacat** [comment]
@yutakahirano Yes with your code above I get ~100MB increase in memory use for 100k `pump()`s if there's no Uint8Array allocations in the "loop." (On Chrome Canary OS X, with GC manually forced.)

**Comment by yutakahirano** [comment]
Thanks @ithinkihaveacat, I will investigate more.

**Comment by tyoshino** [comment]
Created a public version of the issue at http://crbug.com/477661

**Comment by domenic** [comment]
> I ran the script without Uint8Array, and succeeded to reproduce the memory blow-up. I can reproduce the issue with the standalone d8 (V8 shell).

Can you share the full d8-runnable script?

**Comment by yutakahirano** [comment]
```
var N = 1000 * 1000;
var counter = 0;
var resolve;

function read() {
  return Promise.resolve({done: ++counter >= N});
}

function pump() {
  return read().then(function(r) {
    if (r.done) {
      // console.log('done');
      // print('done');

      // 1: Return undefined.
      return;

      // 2. Return a pending promise.
      return new Promise(function() {});

      // 3. Return a pending promise with exposing the resolver.
      return new Promise(function(r) {resolve = r;});

      // 4. continue pumping.
      counter = 0;
      return pump();

      // 5. continue pumping without chaining.
      counter = 0;
      pump();
      return;
    }
    return pump();
  });
}

var p = pump();
```

**Comment by yutakahirano** [comment]
Let me describe the situation.

Let `p1` and `p2` be promises resolved with undefined.
Let `p3` be a pending promise and `resolve` be its resolver function.
Let `p` be `p1.then(() => p2.then(() => p3))`.
In the definition of `p`, We call `p2.then(() => p3)` as `q`.

When we call `resolve`, it resolves `p3` and it resolves `q` and it resolves `p`. That means the system retains those three promises while we aren't interested in `q`. The longer the chain is, the more redundant promises we have (e.g. `p = p1.then(() => p2.then(() => p3.then(() => p4)))`).

I read [the V8 code](https://code.google.com/p/chromium/codesearch#chromium/src/v8/src/promise.js&q=promise.js&sq=package:chromium&type=cs) and checked that it is implemented as described above. I haven't read Mozilla implementation, but I expect that it's similar as it behaves alike for [the example code](https://github.com/domenic/streams-demo/issues/4#issuecomment-93767478).

**Comment by domenic** [comment]
This is sounding more and more like https://github.com/promises-aplus/promises-spec/issues/179. See especially recent discussions at https://github.com/promises-aplus/promises-spec/issues/179#issuecomment-93420260

I will confirm and investigate possible spec fixes.

**Comment by domenic** [comment]
Yes, it is related.

Here is some stuff I learned while drawing [diagrams](https://cloud.githubusercontent.com/assets/617481/7194427/ffeb71c4-e47c-11e4-8b2a-7896d267e499.JPG) of your situation. To recap it is

``` js
const p1 = Promise.resolve();
const p2 = Promise.resolve();

let resolve;
const p3 = new Promise(r => resolve = r);

const p = p1.then(() => {
  const q = p2.then(() => p3);
  return q;
});

// until someone calls resolve(), we hold a reference to q, somewhat unexpectedly
```

The resulting reference graph is p3 -> q -> p. As you mention further recursive levels will add more intermediate nodes.

The reason for these references is the same in both cases. To outline the p3 -> q case, it comes from
- p2 fulfills and executes `() => p3`
- It uses the result, p3, to do resolveQ(p3)
- This quickly results in p3.then(resolveQ, rejectQ)
- So now p3@[[FulfillReactions]] contains { [[Capabilities]]: resultCapabilitiesForQ, [[Handler]]: resolveQ }. (Note that resultCapabilitiesForQ is unused since the return value of .then in this case is ignored. But it doesn't matter as resolveQ -> q, so we anyway have p3 -> q.)

So in a bit more detail the references are p3 -> resolveQ -> q and q -> resolveP -> p.

As @ForbesLindesay points out in the Promises/A+ issue, the analogy to tail recursion is pretty striking, i.e. keeping around intermediate "stacks" (or promises in this case) that are just references to a parent. And you can break it just as easily using techniques like @tyoshino's. So maybe this is OK and we just say that, like a growing synchronous stack for recursive synchronous function calls, you have a growing asynchronous "stack" for recursive asynchronous function calls.

There are optimizations we could do, maybe without changing observable semantics. However the general direction of such optimizations is to move away from resolveQ(p3) meaning p3.then(resolveQ, rejectQ). For example see https://github.com/domenic/cancelable-promise/pull/1. That is bad because it means promise subclasses' `.then` methods, or other thenables, or overwritten then methods on individual promises or on the prototype, are not consulted. We could condition the optimization on `then === originalPromisePrototypeThen` which would sidestep the issue, but would also mean that the "leak" comes back when any of those scenarios occur (including e.g. using an official `CancelablePromise`).

There may be something even more clever we can do. I am vaguely thinking along the lines of some post-processing step that recognizes the { [[Capabilities]]: resultCapabilitiesForQ, [[Handler]]: resolveQ } entry and replaces it with { [[Capabilities]]: undefined, [[Handler]]: resolveP }. But I think that requires weak refs which I'm told will cause a lot of performance overhead of their own.

My overall feeling is that this is troubling. I can kind of say that the status quo is OK because it is like sync recursion, but I would still prefer to solve this. I just am not sure how to do that while also maintaining the interoperability goals of promises with other thenables.

**Comment by tyoshino** [comment]
Thanks for great summary and analysis, Domenic, Yutaka and Michael.

So for now, shall we change the demo to do pump like https://github.com/domenic/streams-demo/issues/4#issue-68686793 ?

**Comment by yutakahirano** [comment]
> There are optimizations we could do, maybe without changing observable semantics. However the general direction of such optimizations is to move away from resolveQ(p3) meaning p3.then(resolveQ, rejectQ). For example see domenic/cancelable-promise#1. That is bad because it means promise subclasses' .then methods, or other thenables, or overwritten then methods on individual promises or on the prototype, are not consulted. We could condition the optimization on then === originalPromisePrototypeThen which would sidestep the issue, but would also mean that the "leak" comes back when any of those scenarios occur (including e.g. using an official CancelablePromise).

Domenic, I have a question. In the following code, Should 'then called' be printed?

```
var y = Promise.resolve(80);
y.then = function() {
  console.log('then called');
  return Promise.prototype.then.apply(y, arguments);
};

var x = Promise.resolve(70);
var z = x.then(function() {
  return y;
});

z.then(function() {
  console.log('z is fulfilled');
});
```

Firefox prints and Chrome doesn't.
The "then" method in your comment is `y`'s "then" method in this example, right?

**Comment by yutakahirano** [comment]
> Should 'then called' be printed?

Seems Chrome is wrong. `x.then(function() {return y;})` calls `PerformPromiseThen(x, function() {return y;}, undefined, capability)` and it enqueues a job calling _Promise Resolve Function_ with `y`. In the function, an job calling `y.then` function is enqueued.

**Comment by domenic** [comment]
@tyoshino yeah I guess we can change the demo with a note that we're hoping this won't be necessary in the future.

@yutakahirano indeed, `then` should be called; V8 is wrong (probably because of their non-standard `accept` semantics).

**Comment by yutakahirano** [comment]
@domenic Thanks! I filed a bug https://code.google.com/p/chromium/issues/detail?id=477921.

**Comment by wanderview** [comment]
> I can kind of say that the status quo is OK because it is like sync recursion, but I would still prefer to solve this.

I'm not sure I agree with this statement.  One normally does not write a sync loop that runs for potentially "infinite" time.

Using Promises in an API like Streams, though, directly encourages building these kind of long running async loops.  This feels like a pretty big footgun to be baking into the API.

(Note, I don't think the Promises in ServiceWorker, fetch, and Cache are a problem because they are not loop oriented APIs.)

I guess what I'm saying is it would be really nice to have a plan/solution for this before finalizing the Streams API.

Also, can you please confirm for me that this callback style async read does not have the same problem?

```
myStream.read(function handleChunk(chunk) {
  processChunk(chunk);
  myStream.read(handleChunk);
});
```

I don't think it does, but I could easily be confused here.

**Comment by wanderview** [comment]
I wrote a gist to compare the promise-vs-callback APIs to try to understand the problem:

https://gist.github.com/wanderview/16f2839ba57514a625c4

**Comment by domenic** [comment]
@wanderview yeah that helps clarify why I am not really OK with this situation even though there is an analogy. Thanks.

I added a note to your gist with another promise version that does not have this problem. The issue is essentially the way in which we propagate the final completion up the chain. Similarly you can come up with callback versions that do have the problem---e.g. the version in your post would have the problem if many chunks were available synchronously, and I could imagine variations using one of the popular async libraries would have the problem if those libraries also try to do chaining of result values in some way.

Anyway I think the candidate fix at https://github.com/domenic/cancelable-promise/pull/1 or something derived from it might be what's necessary here. I hope there's something better which preserves the calls-`.then` property, and will spend a bunch of time this weekend with a whiteboard trying to see if I can come up with anything. But yeah, we need to fix this issue with promises.

**Comment by domenic** [comment]
Interestingly async/await would not have this problem. (Similar to how loops do not have the same stack-overflow problem of recursive functions.)

**Comment by tyoshino** [comment]
I agree with @wanderview about this point:

> I'm not sure I agree with this statement. One normally does not write a sync loop that runs for potentially "infinite" time.

Unless we specify reduction of promise chain as mandatory in the spec as well as tail recursion optimization is required in the Scheme spec, we shouldn't recommend promise-based final result propagation for consuming a stream as idiomatic.

---

As @domenic suggested in https://gist.github.com/wanderview/16f2839ba57514a625c4#comment-1435869, which is equivalent to my code in the OP of this issue, it's not a problem as far as we don't use the chain to propagate the final result.

Even https://gist.github.com/wanderview/16f2839ba57514a625c4#file-callback-async-loop should work if the callback gets run as a microtask.

**Comment by tyoshino** [comment]
> Interestingly async/await would not have this problem. (Similar to how loops do not have the same stack-overflow problem of recursive functions.)

Where did you use async/await? For each step?

**Comment by domenic** [comment]
@yutakahirano I cannot reproduce your results. Running the following in Chrome Canary (44.0.2375.0):

``` html
<!DOCTYPE html>
<meta charset="utf8">

<script>
"use strict";
(function () {
  var N = 1000 * 1000;
  var counter = 0;
  var resolve;

  function read() {
    return Promise.resolve({done: ++counter >= N});
  }

  function pump() {
    return read().then(function(r) {
      if (r.done) {
        return;
      }
      return pump();
    });
  }

  window.p = pump();
}());
</script>
```

by clicking "record" in the timeline, then reloading, then waiting for the spinner to stop, then clicking "collect garbage," then clicking stop recording, I get the following result:

![screenshot](https://cloud.githubusercontent.com/assets/617481/7241006/af63b25c-e783-11e4-9a04-8dd690e57a42.png)

The heap snapshot also reveals only two promises objects alive.

**Comment by yutakahirano** [comment]
@domenic, the root cause is that |resolve| is not exposed to the global environment. Thus the entire chain can be collected once pumping is done.

**Comment by yutakahirano** [comment]
Sorry for the poor English. I updated [the above comment](https://github.com/domenic/streams-demo/issues/4#issuecomment-94606071).

**Comment by domenic** [comment]
@yutakahirano hmm OK, so what is a better test to run? (Hopefully one that parallels the demo?)

**Comment by yutakahirano** [comment]
The original problem was that the demo used much memory. That can be observed through [your code](https://github.com/domenic/streams-demo/issues/4#issuecomment-94575501): it uses a lot of memory and that's not we expected before.

You can make the chain uncollectable by exposing `resolve` to the global environment, though.

```
<!DOCTYPE html>
<meta charset="utf8">

<script>
"use strict";
var resolve;
(function () {
  var N = 1000 * 1000;
  var counter = 0;

...
```

Does that answer your question?

**Comment by domenic** [comment]
@yutakahirano I'm still not clear. My code uses no real non-GCable memory. I assume GC-able memory is not really under discussion.

As for exposing resolve to the global environment, my code doesn't assign `resolve` to anything, so I don't think your modification gives enough of a difference. [Your code](https://github.com/domenic/streams-demo/issues/4#issuecomment-93767478) which I based it off of has some cases where `resolve` is assigned, but by default it just returns undefined so I tried to adapt that.

**Comment by yutakahirano** [comment]
@domenic But the system cannot collect memory while pumping. That's problematic. If you set a larger number to `N`, you will see OOM (or GC thrashing).

**Comment by domenic** [comment]
Yeah but that's because in the example code it's entirely synchronous (since it only goes back to the microtask queue). It gives the engine no time to GC. The demo at least has some delay between chunks (perhaps less delay on slower computers).

If the problem here is really about GC-able memory then this is just a quality of implementation problem and Chrome needs to GC more.

**Comment by yutakahirano** [comment]
> Yeah but that's because in the example code it's entirely synchronous (since it only goes back to the microtask queue). It gives the engine no time to GC.

I don't fully understand the GC in V8 and you may be right in the sense that the memory pressure is partially because of the synchronous style.

> The demo at least has some delay between chunks (perhaps less delay on slower computers).

But I think my point 

> the system cannot collect memory while pumping

is still true. While reading from the network, the chain is growing and the system cannot collect it.

The following `read()` causes OOM.

```
function read() {
  if (++counter % 10000 === 0) {
    return new Promise(function(r) {
      setTimeout(function() { r({done: false}); }, 10);
    });
  }
  return Promise.resolve({done: false});
}
```
